### PR TITLE
`range` uses `TwicePrecision` when possible (part 2)

### PR DESCRIPTION
**Pull request by jipolanco**
Follow-up of #44313. See also discussion in #43059.

Fixes the return type of:

- `range(start::IEEEFloat; step::Real, length)`
- `range(; stop::IEEEFloat, step::Real, length)`
- `range(; stop::Real, step::IEEEFloat, length)`

making sure that `TwicePrecision` is always used in these cases.

### CODE REVIEW

File: base/twiceprecision.jl [hunk 1]
```diff
@@ -456,6 +456,9 @@ step(r::StepRangeLen{T,TwicePrecision{T},TwicePrecision{T}}) where {T} = T(r.ste
 range_start_step_length(a, st::IEEEFloat, len::Integer) =
     range_start_step_length(oftype(st, a), st, len)
 
+range_start_step_length(a::IEEEFloat, st, len::Integer) =
+    range_start_step_length(a, oftype(a, st), len)
```
Comment (vtjnash) [inline base/twiceprecision.jl:460]: We might want to use `promote` for these?
```suggestion
    range_start_step_length(promote(a, st)..., len)
```

**Review by vtjnash — COMMENTED** [review]
Can you add some tests for mixed-IEEEFloat calls, since I think there may be some ambiguity errors there:
```
julia> range(0.0f0, step=1.0e0, length=10)
0.0:1.0:9.0

julia> typeof(ans)
StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}

julia> range(0.0e0, step=1.0f0, length=10)
0.0f0:1.0f0:9.0f0

julia> typeof(ans)
StepRangeLen{Float32, Float64, Float64, Int64}
```

File: base/twiceprecision.jl [hunk 2]
```diff
@@ -456,6 +456,9 @@ step(r::StepRangeLen{T,TwicePrecision{T},TwicePrecision{T}}) where {T} = T(r.ste
 range_start_step_length(a, st::IEEEFloat, len::Integer) =
     range_start_step_length(oftype(st, a), st, len)
 
+range_start_step_length(a::IEEEFloat, st, len::Integer) =
```
Comment (vtjnash) [inline base/twiceprecision.jl:459]:
```suggestion
range_start_step_length(a::IEEEFloat, st::Real, len::Integer) =
```

Should we try to limit this to Reals? Looking back at the old ranges.jl code, that seemed to be the intent. I think we don't want to end up trying to catch this case, for example:
```julia
julia> range(0.0, step=1im, length=10) |> typeof
StepRangeLen{ComplexF64, Float64, Complex{Int64}, Int64}
```

Comment (jipolanco) [reply base/twiceprecision.jl:459]: Makes sense, thanks!

**Comment by jipolanco** [comment]
Now ambiguities should be resolved, unless I missed something. I added a bunch of different tests to check this.

TwicePrecision is also now limited to real-valued ranges, so that e.g. complex ranges work correctly again.

For now all of this is done by defining several variants of `range_start_step_length` and `range_step_stop_length`. I'm sure there's a more elegant way of doing this, but this solution seems to work and is quite simple to understand.

File: base/twiceprecision.jl [hunk 3]
```diff
@@ -453,9 +453,15 @@ end
 step(r::StepRangeLen{T,TwicePrecision{T},TwicePrecision{T}}) where {T<:AbstractFloat} = T(r.step)
 step(r::StepRangeLen{T,TwicePrecision{T},TwicePrecision{T}}) where {T} = T(r.step)
 
-range_start_step_length(a, st::IEEEFloat, len::Integer) =
+range_start_step_length(a::Real, st::IEEEFloat, len::Integer) =
     range_start_step_length(oftype(st, a), st, len)
```
Comment (vtjnash) [inline base/twiceprecision.jl:457]:
```suggestion
    range_start_step_length(promote(a, st)..., len)
```

File: base/twiceprecision.jl [hunk 4]
```diff
@@ -474,7 +480,13 @@ function range_start_step_length(a::T, st::T, len::Integer) where T<:IEEEFloat
     steprangelen_hp(T, a, st, 0, len, 1)
 end
 
-function range_step_stop_length(step::IEEEFloat, stop, len::Integer)
+range_step_stop_length(step::Real, stop::IEEEFloat, len::Integer) =
+    range_step_stop_length(oftype(stop, step), stop, len)
```
Comment (vtjnash) [inline base/twiceprecision.jl:484]:
```suggestion
    range_step_stop_length(promote(step, stop)..., len)
```

File: base/twiceprecision.jl [hunk 5]
```diff
@@ -474,7 +480,13 @@ function range_start_step_length(a::T, st::T, len::Integer) where T<:IEEEFloat
...
+
+range_step_stop_length(step::IEEEFloat, stop::Real, len::Integer) =
+    range_step_stop_length(step, oftype(step, stop), len)
```
Comment (vtjnash) [inline base/twiceprecision.jl:487]:
```suggestion
    range_step_stop_length(promote(step, stop)..., len)
```

**Review by vtjnash — COMMENTED** [review]
Great work! Thank you

**Pull request merged**